File: euler_086.py
from time import time
from math import ceil, floor, sqrt

def is_square(n):
    return int(sqrt(n))**2 == n

def number_of_int_sols(M):
    count = 0
    for z in range(3, M + 1):
        for xy in range(1, 2 * z + 1):
            if is_square(z ** 2 + xy ** 2):
                if xy <= z:
                    count += floor(xy / 2)
                else:
                    count += 1 + z - ceil(xy / 2)
    return count

start = time()

M = 1000
while number_of_int_sols(M) < 1000000:
    M += 100

print(M, number_of_int_sols(M), M - 100, number_of_int_sols(M - 100))

M = 1800
while number_of_int_sols(M) < 1000000:
    M += 10

print(M, number_of_int_sols(M), M - 10, number_of_int_sols(M - 10))

M = 1810
while number_of_int_sols(M) < 1000000:
    M += 1

print(M, number_of_int_sols(M), M - 1, number_of_int_sols(M - 1))

stop = time()
print("Time: " + str(stop - start))

File: euler_086.js
const start = Date.now();

// Box dimensions: a x b x c, a >= b >= c
const shortestPath = (a,b,c) => {
    const p1 = Math.sqrt((a + c)**2 + b**2);
    const p2 = Math.sqrt((b + c)**2 + a**2);
    const m = Math.min(p1, p2);
    if (Number.isInteger(m)) {
        return m;
    }
    return 0;
}

const numberOfIntegerPaths = (M) => {
    let count = 0;
    for (let a = 1; a < M + 1; a++) {
        for (let b = 1; b < a + 1; b++) {
            for (let c = 1; c < b + 1; c++) {
                if (shortestPath(a,b,c) > 0) {
                    count++;
                }
            }
        }
    }
    return count;
}

// Closest 100 is 18
// Closest 10 is 1
let ones = 1;
while (numberOfIntegerPaths(18 * 100 + 1 * 10 + ones) < 1000000) {
    ones++;
}
ones--; // ones = 7

const end = Date.now();

console.log("Execute time: ", end - start, " The least value of M is: ", 1817);